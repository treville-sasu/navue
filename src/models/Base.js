export class Model {
  // eslint-disable-next-line no-unused-vars
  constructor({ type, ...properties } = {}) {
    Object.assign(this, properties);
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  // TODO write specs
  clone() {
    return this.constructor.from(JSON.parse(JSON.stringify(this)));
  }

  static from({ type, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) return new this(properties);
    else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}

// TODO : implementation for when Vuejs Suppots Array-like objects.
// export class Store extends Array {
export class Store {
  constructor(properties = {}, ...items) {
    this.items = items;
    Object.assign(this, properties);
  }

  // get items() {
  //   return [...this];
  // }

  [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  get length() {
    return this.items.length;
  }

  valueOf() {
    return this.items;
  }

  get properties() {
    // eslint-disable-next-line no-unused-vars
    let { items, ...props } = { ...this };
    return props;
    // return Object.fromEntries(
    //   Object.entries(this).filter(([key]) => isNaN(key))
    // );
  }

  get first() {
    return this.items[0];
  }

  get last() {
    return this.items[this.length - 1];
  }

  add(item, index) {
    index ? this.items.splice(index, 0, item) : this.items.push(item);
    return this.length;
  }

  remove(item, index) {
    if (index === undefined) index = this.items.indexOf(item);
    return index >= 0 ? this.items.splice(index, 1).length == 1 : false;
  }

  keep(filter) {
    let i = this.items.length;
    while (i--) {
      let result = filter.call(this.items, this.items[i]);

      if (result == undefined) this.items.splice(i, 1);
      else if (!Object.is(result, this.items[i]))
        this.items.splice(i, 1, result);
    }
    return this;
  }

  toJSON() {
    return {
      ...this.properties,
      items: this.items,
      type: this.constructor.name
    };
  }

  static from({ type, items = [], ...properties }, ...constructors) {
    if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      let constructor = constructors.shift();
      return new this(
        properties,
        ...items.map(i => {
          return constructor ? constructor.from(i, ...constructors) : i;
        })
      );
    }
  }
}
