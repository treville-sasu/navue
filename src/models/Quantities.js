import { Model } from "@/models/Base.js";

export class Quantity extends Model {
  // eslint-disable-next-line no-unused-vars
  constructor(value, _unit, { type, unit, precision, ...properties } = {}) {
    if (arguments[0] instanceof Quantity) return arguments[0];
    else {
      super(properties);

      if (_unit) this.unit = _unit;
      else this._unit = undefined;

      if (value || value === 0) this.value = value;
      else {
        this._value = undefined;
        this.precision = undefined;
      }

      if (precision) this.precision = precision;

      if (unit) this.unit = unit;
    }
  }

  get value() {
    return this.constructor.significantValue(
      this._to(this._value),
      this.precision
    );
  }

  set value(val) {
    let f = Number(val);
    if (Number.isNaN(f)) throw `\`${val}\` could not be coerced to a number`;
    else {
      this._value = this._from(f);
      this.precision = Math.max(3, this.constructor.significantFigures(val));
    }
  }

  get unit() {
    return this._unit || this.constructor.baseUnit;
  }

  set unit(val) {
    if (!this.constructor.units[val]) throw `'${val}' is not an available unit`;
    else this._unit = val;
  }

  get factor() {
    return this.constructor.units[this.unit];
  }

  _from(val, unit = this.unit) {
    return Number(val) / this._factor(unit);
  }

  _to(val, unit = this.unit) {
    return Number(val) * this._factor(unit);
  }

  _factor(unit) {
    if (this.constructor.units[unit]) return this.constructor.units[unit];
    else throw `'${unit}' is not an available unit`;
  }

  to(unit) {
    return this._to(this.value, unit);
  }

  as(unit, precision) {
    return Object.assign(this.constructor.from(this), { unit, precision });
  }

  add(other, props) {
    if (typeof other == "number" || other instanceof this.constructor)
      return new this.constructor(this + other, undefined, {
        precision: Math.min(this.precision, other.precision),
        unit: this._unit,
        ...props
      });
    else throw "operands should be of the same type";
  }

  // FIXME : prod with quantity should not return the type from this nor other.
  prod(other, props) {
    return new this.constructor(this * other, undefined, {
      precision: Math.min(
        this.precision,
        other.precision || this.constructor.significantFigures(other)
      ),
      unit: this._unit,
      ...props
    });
  }

  valueOf() {
    return this._value;
  }

  toString() {
    let sf = Math.max(1, Math.min(this.precision, 21));
    return `${
      isNaN(this.value)
        ? "-"
        : this.value.toLocaleString(undefined, {
            minimumSignificantDigits: sf,
            maximumSignificantDigits: sf
          })
    } ${this.unit}`;
  }

  toJSON() {
    // eslint-disable-next-line no-unused-vars
    let { _value, _unit, ...properties } = this;
    return {
      ...properties,
      value: this.value,
      unit: this.unit,
      type: this.constructor.name
    };
  }

  static significantFigures(n) {
    let s;
    if (n === 0) {
      return 0;
    } else if (typeof n === "number") {
      s = Math.abs(n).toExponential();
    } else if (typeof n === "string" && !Number.isNaN(Number(n))) {
      s = n.replace(/^\+|-/, "").replace(/^0+/, "");
    } else return;

    s = s.replace(".", "");
    const i = s.lastIndexOf("e");
    return i > 0 ? i : s.length;
  }

  static significantValue(val, precision = 1) {
    if (val === 0) return 0;
    const fact = Math.pow(
      10,
      precision - Math.floor(Math.log(val < 0 ? -val : val) / Math.LN10) - 1
    );
    return Math.round(val * fact) / fact;
  }

  static get units() {
    throw "list of available units should be set on constructor.";
  }

  static get baseUnit() {
    return Object.keys(this.units)[Object.values(this.units).indexOf(1)];
  }

  static from({ type, value, unit, ...properties } = {}) {
    if (!type || this.name == type) return new this(value, unit, properties);
  }
}

export class Distance extends Quantity {
  static get units() {
    return {
      m: 1,
      Km: 0.001,
      ft: 3.28084,
      NM: 0.000539957,
      mi: 0.000621371,
      yd: 1.09361
    };
  }
}

export class Altitude extends Distance {
  constructor(value, unit, { reference, ...properties } = {}) {
    if (unit == "FL") reference = "QNE";
    else if (unit && !reference)
      throw `property 'reference' should be given, got '${reference}'`;

    super(value, unit, { ...properties, reference });
  }

  // densityAltitude(temp) {
  //   return new this.constructor(
  //     this.value + 36.576 * (temp - 15),
  //     "m",
  //     "MSL"
  //   );
  // }

  get value() {
    if (this.unit == "FL")
      return this.constructor.significantValue(this._to(this._value), 2);
    else return this._to(this._value);
  }

  set value(val) {
    super.value = val;
  }

  toString() {
    if (this.unit == "FL") {
      return `FL${String(this.value).padStart(3, "0")}`;
    } else if (this.value && this.reference) {
      return `${super.toString()} ${this.reference}`;
    }
  }

  static get references() {
    return ["MSL", "AGL", "QNH", "QNE", "QFE", "WGS84"];
  }

  static get units() {
    return {
      m: 1,
      ft: 3.28084,
      FL: 0.0328084
    };
  }
}

export class Speed extends Quantity {
  static get units() {
    return {
      "m/s": 1,
      "km/h": 3.6,
      "ft/min": 196.85,
      kt: 1.94384,
      mph: 2.23694
    };
  }

  // TAS(altitude, temperature) {}
}

export class Volume extends Quantity {
  static get units() {
    return {
      m3: 1,
      L: 1000,
      "gal US": 264.172,
      "gal Imp": 219.969
    };
  }
}

export class Consumption extends Volume {
  static get references() {
    return ["h", "1000ft", "u"];
  }

  toString(precision) {
    return `${super.toString(precision)}/${this.reference}`;
  }
}

export class Weight extends Quantity {
  static get units() {
    return {
      kg: 1,
      lb: 2.20462,
      t: 0.001,
      "L-avgas": 1 / 0.721,
      "L-JetA1": 1 / 0.8
    };
  }
}

export class Angle extends Quantity {
  static get units() {
    return {
      rad: 1,
      "°": 180 / Math.PI,
      gon: 200 / Math.PI,
      rev: 1 / 2 / Math.PI,
      "₥": 6400 / 2 / Math.PI
    };
  }
}

export class Azimuth extends Angle {
  get value() {
    //FIXME: 0 should give 360
    return this._to((this._value + 2 * Math.PI) % (2 * Math.PI));
  }

  set value(val) {
    super.value = val;
  }
}

export class Bearing extends Angle {
  get value() {
    //FIXME: 180 % should give 180
    return this._to(((this._value + 3 * Math.PI) % (2 * Math.PI)) - Math.PI);
  }

  set value(val) {
    super.value = val;
  }
}

export class Pressure extends Quantity {
  static get units() {
    return {
      Pa: 1,
      hPa: 0.01,
      Psi: 1 / 6894.76,
      atm: 1 / 101325,
      bar: 0.00001,
      inHg: 1 / 3386.389
    };
  }
}

// TODO Spec it !
export class Temperature extends Quantity {
  static get units() {
    return {
      "°C": [1, 0],
      K: [1, 273.15],
      "°F": [9 / 5.32]
    };
  }

  _from(val, unit = this.unit) {
    return Number(val) / this._factor(unit)[0] + this._factor(unit)[1];
  }

  _to(val, unit = this.unit) {
    return (Number(val) - this._factor(unit)[1]) * this._factor(unit)[0];
  }
}
export class Frequency extends Quantity {
  static get units() {
    return {
      Hz: 1,
      KHz: 0.001,
      MHz: 0.00001,
      GHz: 0.00000001
    };
  }
}
