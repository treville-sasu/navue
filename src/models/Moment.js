import { Model } from "@/models/Base.js";
import { Weight, Distance } from "@/models/Quantities.js";

export class Moment extends Model {
  constructor(mass, lever, properties = {}) {
    super({
      mass: new Weight(mass),
      lever: new Distance(lever),
      ...properties
    });
  }

  get value() {
    return this.mass.value * this.lever.value;
  }

  get _value() {
    return this.mass * this.lever;
  }

  valueOf() {
    return this._value;
  }

  get unit() {
    return `${this.mass.unit}.${this.lever.unit}`;
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  static from({ type, mass, lever, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) {
      if (mass) mass = Weight.from(mass);
      if (lever) lever = Distance.from(lever);
      return new this(mass, lever, properties);
    } else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }

  static linearCoG(...moments) {
    let result = moments.flat().reduce((acc, { mass, lever }) => {
      acc.mass = acc.mass.add(mass);
      acc.lever = acc.lever.add(mass * lever);
      return acc;
    }, new Moment(0, 0));

    result.lever = result.lever.prod(1 / result.mass);
    return result;
  }
}
