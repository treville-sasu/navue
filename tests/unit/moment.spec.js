/* eslint-disable jest/no-hooks */
import { Model } from "@/models/Base.js";
import { Distance, Weight } from "@/models/Quantities.js";
import { Moment } from "@/models/Moment.js";

describe("moment", () => {
  let moment;
  beforeAll(() => {
    moment = new Moment(new Weight(1 / 2), new Distance(2));
  });

  it("extends Model", () => {
    expect(Moment.prototype).toBeInstanceOf(Model);
  });

  it("got a lever", () => {
    expect(moment.lever).toBeInstanceOf(Distance);
  });

  it("got a mass", () => {
    expect(moment.mass).toBeInstanceOf(Weight);
  });

  it("got a value", () => {
    expect(moment).toHaveProperty("value");
    expect(moment.value).toBeCloseTo(1);
    expect(Number(moment)).toBeCloseTo(1);
  });

  it("got a unit", () => {
    expect(moment).toHaveProperty("unit");
    expect(moment.unit).toMatch(moment.mass.unit);
    expect(moment.unit).toMatch(moment.lever.unit);
  });

  it("export toJSON", () => {
    expect(moment).toHaveProperty("toJSON");
    expect(moment.toJSON()).toMatchSnapshot();
  });

  it("import from literal", () => {
    expect(
      Moment.from({
        lever: {
          precision: 1,
          type: "Distance",
          unit: "m",
          value: 2
        },
        mass: {
          precision: 1,
          type: "Weight",
          unit: "kg",
          value: 0.5
        },
        type: "Moment"
      })
    ).toMatchSnapshot();
  });

  it("linearCoG(), creates moment from 1D lever & mass arrray", () => {
    expect(
      Moment.linearCoG({ mass: 1 / 2, lever: 2 }, { mass: 1 / 2, lever: -2 })
    ).toStrictEqual(new Moment(1, 0));
    expect(
      Moment.linearCoG({ mass: 3, lever: 2 }, { mass: 2, lever: -2 })
    ).toStrictEqual(new Moment(5, 0.4));
    expect(
      Moment.linearCoG({ mass: 1, lever: 1 }, { mass: 1, lever: -1 })
    ).toStrictEqual(new Moment(2, 0));
  });
});
