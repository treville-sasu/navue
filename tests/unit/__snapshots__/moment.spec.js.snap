// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moment export toJSON 1`] = `
Object {
  "lever": Object {
    "precision": 1,
    "type": "Distance",
    "unit": "m",
    "value": 2,
  },
  "mass": Object {
    "precision": 1,
    "type": "Weight",
    "unit": "kg",
    "value": 0.5,
  },
  "type": "Moment",
}
`;

exports[`moment import from literal 1`] = `
Object {
  "lever": Object {
    "precision": 1,
    "type": "Distance",
    "unit": "m",
    "value": 2,
  },
  "mass": Object {
    "precision": 1,
    "type": "Weight",
    "unit": "kg",
    "value": 0.5,
  },
  "type": "Moment",
}
`;
